internal/faketime/faketime.go

```diff
@@ -34,32 +35,28 @@ func AutoAdvance(t time.Time, dt time.Duration) func() time.Time {
 // TimeAdvance allows controlling the passage of time. Intended to be used in
 // tests.
 type TimeAdvance struct {
-	mu sync.Mutex
-	t  time.Time
+	base  time.Time
+	delta time.Duration
 }
 
 // NewTimeAdvance creates a TimeAdvance with the given start time
 func NewTimeAdvance(start time.Time) *TimeAdvance {
-	return &TimeAdvance{t: start}
+	return &TimeAdvance{base: start}
 }
 
 // NowFunc returns a time provider function for t
 func (t *TimeAdvance) NowFunc() func() time.Time {
 	return func() time.Time {
-		t.mu.Lock()
-		defer t.mu.Unlock()
+		dt := time.Duration(atomic.LoadInt64((*int64)(&t.delta)))
 
-		return t.t
+		return t.base.Add(dt)
 	}
 }
 
 // Advance advances t by dt, such that the next call to t.NowFunc()() returns
 // current t + dt
 func (t *TimeAdvance) Advance(dt time.Duration) time.Time {
-	t.mu.Lock()
-	defer t.mu.Unlock()
+	advance := atomic.AddInt64((*int64)(&t.delta), int64(dt))
```

Comment:
i'm actually surprised this works (casting to *int64).

<optional>
in my comment I was thinking about keeping two numbers, which would be int64.

Something like:

```
type TimeAdvance struct {
  timestampNanoseconds int64
  advanceNanoseconds int64
}
```

then use `time.Unix(0, atomic.LoadInt64(&t.timestampNanoseconds))`
</optional>

Reply:
This can be simplified so some of the casts can be removed.